parsely/parselyandroid/ParselyTracker.java

```diff
@@ -521,4 +611,137 @@ protected void onPostExecute(String advertId) {
+     * Implemented to handle its own queuing of future executions to accomplish
+     * two things:
+     *
+     * 1. Flushing any engaged time before canceling.
+     * 2. Progressive backoff for long engagements to save data.
+     */
+    private class EngagementManager {
+
+        public Map<String, Object> baseEvent;
+        private boolean started;
+        private Timer parentTimer;
+        private TimerTask waitingTimerTask;
+        private long latestDelayMillis, totalTime;
+
+
+        public EngagementManager(Timer parentTimer, long intervalMillis, Map<String, Object> baseEvent) {
+            this.baseEvent = baseEvent;
+            this.parentTimer = parentTimer;
+            this.latestDelayMillis = intervalMillis;
+            this.totalTime = 0;
+        }
+
+        public boolean isRunning() { return this.started; }
+
+        public void start() {
+            this.scheduleNextExecution(this.latestDelayMillis);
+            this.started = true;
+        }
+
+        public void stop() {
+            this.waitingTimerTask.cancel();
+            this.started = false;
+
+        }
+
+        public boolean isSameVideo(String url, ParselyVideoMetadata metadata) {
+            Map<String, Object> baseMetadata = (Map<String, Object>) baseEvent.get("metadata");
+            return (baseEvent.get("url") == url &&
+                    baseMetadata.get("canonical_url") == metadata.canonical_url &&
+                    (int) (baseMetadata.get("duration")) == metadata.durationSeconds);
+        }
+
+        private void scheduleNextExecution(long delay) {
+            TimerTask task = new TimerTask(){
+                public void run(){
+                    doEnqueue(this.scheduledExecutionTime());
+                    updateLatestInterval();
+                    scheduleNextExecution(latestDelayMillis);
+                }
+
+                public boolean cancel() {
+                    doEnqueue(this.scheduledExecutionTime());
+                    return super.cancel();
+                }
+            };
+            PLog(String.format("latestDelayMillis: %d", delay));
+            this.latestDelayMillis = delay;
+            this.parentTimer.schedule(task, delay);
+            this.waitingTimerTask = task;
+        }
+
+        private void doEnqueue(long scheduledExecutionTime) {
+            // Create a copy of the base event to enqueue
+            Map<String, Object> event = new HashMap(this.baseEvent);
+            PLog(String.format("Enqueuing %s event.", event.get("action")));
+
+            // Adjust inc by execution time in case we're late or early.
+            long executionDiff = (System.currentTimeMillis() - scheduledExecutionTime);
+            long inc = (this.latestDelayMillis + executionDiff) / 1000;
+            this.totalTime += inc;
+            event.put("inc", inc);
+            event.put("tt", this.totalTime);
+
+            enqueueEvent(event);
+        }
+
+        private void updateLatestInterval() {
```

Comment:
If this function doesn't care about the `duration` of the thing being tracked, `duration` doesn't need to be required by the public API. Additionally, an automatic backoff of 25% for both engaged time and video is a departure from the behavior of the JS tracker.

Reply:
It is a departure from the current behavior but mirrors things we have in development. I see no reason not to include it, especially since apps are expected to be much more video heavy than what we normally track. It's been tested to not lose information and reduces data we send. Something more important on mobile devices.